packages/block-library/src/cover/shared.js

```diff
@@ -109,3 +117,81 @@ export function getPositionClassName( contentPosition ) {
 
 	return POSITION_CLASSNAMES[ contentPosition ];
 }
+
+/**
+ * Performs a Porter Duff composite source over operation on two rgba colors.
+ *
+ * @see https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcover
+ *
+ * @param {import('colord').RgbaColor} source Source color.
+ * @param {import('colord').RgbaColor} dest   Destination color.
+ * @return {import('colord').RgbaColor} Composite color.
+ */
+function compositeSourceOver( source, dest ) {
+	return {
+		r: source.r * source.a + dest.r * dest.a * ( 1 - source.a ),
+		g: source.g * source.a + dest.g * dest.a * ( 1 - source.a ),
+		b: source.b * source.a + dest.b * dest.a * ( 1 - source.a ),
+		a: source.a + dest.a * ( 1 - source.a ),
+	};
+}
+
+function retrieveFastAverageColor() {
+	if ( ! retrieveFastAverageColor.fastAverageColor ) {
+		retrieveFastAverageColor.fastAverageColor = new FastAverageColor();
+	}
+	return retrieveFastAverageColor.fastAverageColor;
+}
+
+/**
+ * This method evaluates if the cover block's background is dark or not and this boolen
+ * can then be applied to the relevant attribute to help ensure that text is visible by default.
+ * This needs to be recalculated in all of the following Cover block scenarios:
+ * - When an overlay image is added, changed or removed
+ * - When the featured image is selected as the overlay image, or removed from the overlay
+ * - When the overlay color is changed
+ * - When the overlay color is removed
+ * - When the dimRatio is changed
+ *
+ * See the comments below for more details about which aspects take priority when
+ * calculating the relative darkness of the Cover.
+ *
+ * @param {string} url
+ * @param {number} dimRatio
+ * @param {string} overlayColor
+ * @return {boolean|Promise} True if cover should be considered to be dark.
+ */
+export async function getCoverIsDark( url, dimRatio = 50, overlayColor ) {
+	const overlay = colord( overlayColor )
+		.alpha( dimRatio / 100 )
+		.toRgb();
+	// If the dimRatio is less than 50, the image will have the most impact on darkness.
+	if ( url && dimRatio <= 50 ) {
```

Comment:
```suggestion
	if ( url ) {
```

The new heuristic works on all dim ratios.